File: packages/core/rules/test/dos/negative.js
// Test cases that should NOT be detected (false positives)

const fs = require('fs');

// Loop with break condition
while (true) {
  if (shouldStop()) {
    break;
  }
  processData();
}

// Recursion with depth limit
function factorial(n, depth = 0) {
  if (depth > MAX_DEPTH) throw new Error("Max depth");
  if (n <= 1) return 1;
  return n * factorial(n - 1, depth + 1);
}

// File reading with size check
const stats = fs.statSync(userFilePath);
if (stats.size < MAX_FILE_SIZE) {
  const data = fs.readFileSync(userFilePath);
}

// Safe regex
const pattern = /[a-z]+/;
pattern.test(userInput);

// Memory allocation with limit
const size = Math.min(userSize, MAX_SIZE);
const arr = new Array(size);

File: packages/core/rules/test/dos/negative.py
# Test cases that should NOT be detected (false positives)

# Loop with break condition
while True:
    if should_stop():
        break
    process_data()

# File reading with size check
import os
file_stats = os.stat(user_file_path)
if file_stats.st_size < MAX_FILE_SIZE:
    with open(user_file_path) as f:
        content = f.read()

# Chunked file reading
with open(user_file_path) as f:
    for line in f:
        process_line(line)
